Snaps Canary Release

rekmarks opened the pull request:
Closes #11626

rekmarks committed:
Add Snaps

* Implement bls-signer support
* Implement filsnap support
* Add hooknames and hook selectors to RPC methods
* Fix filsnap installation on vault creation
* Various build and CI config updates
* Update sourcemap validation
* Update Sentry configuration
* Change name of production environment to 'filecoin-production'
* Ignore useless sentry error in e2e tests
* Update README.md
* Snaps update 0.0.7 (#11854)
* Add iframe execution environment (#11865)
* Remove WebWorkerExecutionEnvironmentService import
* Ignore filsnap bundle in depcheckrc (#11866)
  * This stops the `depcheck` check from failing.
* Restore Firefox compatibility (#11867)
* Notify user of snap initial execution errors (#11901)
* Add alert on snap initial execution errors
* Changed window.alert to be a browser notification
* Add guard for pluginController removePlugin
* Remove Pre-Installed Filsnap (#12111)
* Add more global snap hooks (#12113)
* Changed snap imports to use `@metamask/` org (#12238)
* Remove all uses of @mm-snaps
* snaps packages @0.3.1 (#12610) (squashed)
* Fence all Flask code
* Remove snap_manageAssets and AssetController
* Add locale messages for expected production permissions
* Remove permissionsDescriptions state (replace by locale messages)
* Remove unnecessary modifications to various files
* Update Flask LavaMoat policy
* Fix permission description display
* Restore LavaMoat beta policy
* Fix permission locale message handling
* Fix build errors, regenerate build system policy
* Fix some broken code fences and translation calls
* Added snaps error notification (#12860)
* skunkworks@0.5.0, migrate snap restricted methods
* Clean up untrusted connection subject type logic
* Record Snap subject metadata
* Fix useOriginMetadata React hook
* Backported bug fix from snaps-0.1 to snaps (#13094)
* Added Snaps Custom Confirmation (#13036)
* Fix bug when adding subject metadata (#13102)

Co-authored-by: Erik Marks <rekmarks@protonmail.com>
Co-authored-by: George Marshall <george.marshall@consensys.net>
Co-authored-by: Hassan Malik <hassan.malik@consensys.net>
Co-authored-by: Olaf Tomalka <olaf.tomalka@gmail.com>
Co-authored-by: Shane Jonas <jonas.shane@gmail.com>

rekmarks commented on ui/components/app/permission-page-container/permission-page-container-content/permission-page-container-content.component.js:
```diff
@@ -33,9 +42,29 @@ export default class PermissionPageContainerContent extends PureComponent {
+        description = t(permissionName);
+      }
+      ///: BEGIN:ONLY_INCLUDE_IN(flask)
+      else if (permissionName.startsWith(invokeSnapPrefix)) {
+        description = t(invokeSnapPrefix, [
+          permissionName.replace(invokeSnapPrefix, ''),
+        ]);
+      } else if (permissionName.startsWith(getBip44EntropyPrefix)) {
+        // TODO:flask create coin_type to protocol name enum
+        description = t(getBip44EntropyPrefix, [
+          permissionName.replace(getBip44EntropyPrefix, ''),
+        ]);
+      } else {
+        description = t(permissionName);
+      }
```
@Gudahtt, I decided to continue using permission names (or, in the case of namespaced permissions, target keys without `*`, since that character apparently cannot appear as top-level object keys in extension JSON files) as the locale message key for permission descriptions, but with this implementation everything except `getBip44Entropy` (which appears as a string literal in this file) is flagged as unused by the locale message verification script. How do you recommend we deal with this?

Gudahtt replied:
That script requires the permission names to be _somewhere_ on disk at build time. That should be the case here too, right?

It would be easiest if we could use string-literals here somehow. We can direct the script to look elsewhere in special cases though.

rekmarks replied:
Yeah, I was thinking I could create a utility function with a switch statement. I'll get to it!

rekmarks replied:
Done here: 406ef67

Gudahtt replied:
Sorry I don't follow how that solves the issue. This still isn't compatible with the strict mode of verify locales, because we aren't passing string literals to the `t` function.

Though it looks like all the strings we need have been laid out now at least. We can switch to using a mapping of enums to messages, so that the linter can find these messages more easily. Like we did here for example: https://github.com/MetaMask/metamask-extension/blob/develop/ui/pages/settings/advanced-tab/advanced-tab.component.js#L453

rekmarks replied:
Ah, I didn't run the strict version of the script locally. I'll get to it!

rekmarks replied:
Although, I have to say that I am not a fan of the strict version of that script, because it's forcing me to use magic strings instead of enums. We should consider creating message enums that can be interpreted by that script so that we can use enums without having to do AST analysis.

Gudahtt commented on app/scripts/controllers/permissions/index.js:
```diff
@@ -1,4 +1,9 @@
 export * from './caveat-mutators';
 export * from './background-api';
+export * from './enums';
```
This change has been extracted to #13012

Gudahtt commented on app/scripts/lib/rpc-method-middleware/createMethodMiddleware.js:
```diff
@@ -1,4 +1,8 @@
+import { handlers as permittedSnapMethods } from '@metamask/rpc-methods/dist/permitted';
+///: END:ONLY_INCLUDE_IN
 import { permissionRpcMethods } from '@metamask/snap-controllers';
+import { selectHooks } from '@metamask/rpc-methods/dist/utils';
```
This change was extracted to #13027

Gudahtt commented on ui/components/app/home-notification/home-notification.stories.js:
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
```
This has been extracted to https://github.com/MetaMask/metamask-extension/pull/13035

rekmarks committed:
Snaps skunkworks@0.6.0 (#13076)

Updates all `snaps-skunkworks` packages to `0.6.0`, and uses the `0.3.0` iframe execution environment.

rekmarks committed:
skunkworks@0.6.1

rekmarks committed:
Add network access endowment permission

rekmarks committed:
Remove unused temporary snap utilities

rekmarks committed:
Add Snaps network access permission (#13119)

Adds the ability for Snaps to access the Internet via the `endowment:network-access` permission. Also adds permission icons, renames permission locale message keys, and fixes Snap permission locale message retrieval. As part of fixing Snap permission locale messages, adds `@metamask/slip44` as a production dependency in order to display the English names of SLIP-44 protocols in our UI.

rekmarks committed:
Make Snaps icons visible in UI (#13120)

We previously added Snap SVG icons to controller state. This PR makes their icons visible in our UI by converting their raw SVG contents to embeddable strings that we use as the `src` property of `img` tags. A new selector, `getTargetSubjectMetadata` is added to support this use case in a memoized fashion.

rekmarks committed:
skunkworks@0.6.2 (#13121)

Updates skunkworks packages to version `0.6.2`. The changes are just minor fixups, including the renaming of a couple of exports and avoiding storing Snap SVG icons in state twice.

rekmarks committed:
@metamask/slip44@2.0.0 (#13122)

Bumps `@metamask/slip44` to `2.0.0`. The changes aren't breaking for our purposes; we just get a longer list of chains.

shanejonas committed:
Changed Snap custom confirmation `approvalText` to "approve" (#13130)

rekmarks committed:
skunkworks@0.6.3

ritave committed:
Added Snap Install page to the UI (#13117)

* Added Snap Install page to the UI

PR review changes

Added flask guards

* Changes after PR review

* Changes after PR review

* Fix locale

* Revert page count visible

* Fix cancel not working when installing snap

* Revert approveLimit locale change

* Fix snaps-confirm.js

Gudahtt commented on ui/selectors/permissions.js:
```diff
@@ -17,16 +18,6 @@ export function getPermissionSubjects(state) {
- * @param {Object} state - The current state.
- * @returns {Object} The permission subjects metadata object.
- */
-export function getPermissionSubjectsMetadata(state) {
```
This change was extracted to https://github.com/MetaMask/metamask-extension/pull/13144

Gudahtt commented on ui/pages/home/home.component.js:
```diff
@@ -232,6 +248,43 @@ export default class Home extends PureComponent {
+                          variant={TYPOGRAPHY.H5}
+                          fontWeight={FONT_WEIGHT.NORMAL}
+                        >
+                          Something Went Wrong
```
This should be localized

Gudahtt replied:
Fixed in #13146

Gudahtt committed:
Localize the snap error messages (#13146)

The messages shown when a snap encounters an error have been localized.
A pre-existing "Something went wrong" message is used, and the message
including the snap error itself was added.

The "Something went wrong" message is a little different in that it
includes the word "Oops!" at the start. This doesn't seem great to me,
but it's minor enough that we can adjust it later if necessary. We can
probably remove the "Oops!" from all places rather than create a second
localized message without it.

Gudahtt committed:
Simplify the `removeSnapError` function (#13145)

The `removeSnapError` function has been changed from an action creator
to a normal function, as the only actions dispatched were unnecessary.
It was dispatching actions to update the `metamask` state, but the
background state will automatically update on its own.

It also uses `async/await` rather than `.then` now, and the callsites
have been updated to use `async/await`. Platforms that support async
stack traces rely upon `async/await` for that enhancement to errors, so
this will result in better stack traces on those platforms.

hmalik88 committed:
Snaps settings page (#13114)

* Add snaps settings page

Fix WIP test-data to use object format

Added snap actions to hook up to background

updated pathname key for snaps view page

added boolean to not render subheader in settings for snaps view page, added settings subheader class to snaps view page

updated locale messages, renamed snap prop to currentSnap, added in title & subtitle and placeholder text for when there are no snaps

added icons folder and included a verified icon for flask

updated snaps test data

updated keys per prettier

made authorship pill self enclosing, added box around pill to introduce padding

added scss file for view-snap and added that to settings page scss file

added isSnapViewPage prop and conditionally display subheader and title

updated story to include backRoute and isSnapViewPage props

fixed spacing

added class to subheader and updated description to use correct color

update test data

updated key for permission div

updated view snap styling

updated view snap to have multiple sections encapsulated by box component

updated locale messages with snapAccess key

added translation call around snap_confirm

added subtext for permissions list section

formatted w/prettier

* linter fixes

* made fixes per comments

* moved components into flask directory and updated imports

* updated remove snap section

* added connected sites list

* lint fixes

* style-lint fix

* changed from snap.name -> snap.id

* fixes per comments, updated test data and used appropriate properties for snaps

* updated dispatch call

* removed unnecessary args

* Fixed some props and action issues with snaps settings page (#13118)

* used idx as key for permission

* added snaps key to messages

* added snaps settings description key

* added code fence

* another code fence

* updated scss, test data, used correct action for onDisconnect

* fixes

* removed memoization in snaps settings card, added action to remove snap permissions from all subjects

* updated padding left and right to be 18px

* added tooltip to settings card

* fixed settings routing issue and added another tooltip to the snaps settings card

* removed padding

* removed condition for settings title

* fixed view-snap page

* style lint fixes

* fixes per comments

* updated actions, moved view snap out of snap-list-tab, updated containers

* style lint fixes

* final fixes

* last fixes

* Remove stray console.log

* Update actions to use convenience function

* Pass `permissionName` to `getSubjectsWithPermission` selector

* Make `removeSnap` a synchronous function

Action creators must be synchronous. The function they return can be
async though.

* Move navigation after removing snap to `useEffect` hook

The UI would crash after removing a snap because the snap metadata
object for the current page was no longer present. Instead a
`useEffect` hook is used to navigate away from the page when the snap
is removed.

This has the added benefit of ensuring the page doesn't crash when the
snap is removed in another tab as well.

* Fix lint error accidentally introduced when rebasing

Co-authored-by: shanejonas <jonas.shane@gmail.com>
Co-authored-by: Mark Stacey <markjstacey@gmail.com>

Gudahtt committed:
Remove unnecessary `!important` (#13148)

This rule was applied just fine without `!important` being present.
Better to avoid using `!important` if we can to make it easier to
understand how these styles are applied, and to allow overriding this
if necessary.

Gudahtt committed:
Fix propType validation warning (#13149)

A propType validation warning was shown when the snap confirmation text
area was rendered because a `value` prop was set without `onChange`
being set, which is often a mistake. In our case we don't want to allow
user interaction with this text area, but we forgot to mark it as read-
only.

The text area is now set as read-only, which silences the error.

ritave committed:
Added metric event on snapInstall message (#13190)

Gudahtt commented:
This branch has just been rebased on `develop`

Gudahtt commented:
Just rebased onto `develop` again

hmalik88 committed:
Snaps UI fixes (#13211)

* QA fixes

* style lint fix

Gudahtt commented:
Rebased again! The auto-reject bug (#13192) should now be fixed.

Gudahtt committed:
Fix lint error introduced by merge conflict

Gudahtt committed:
Replace Sass color variables with CSS variables

Gudahtt committed:
Remove superfluous type definition to fix JSDoc lint error

The JSDoc type definition for the permission specification hooks has
been replaced with a generic type definition for hooks. This was the
wrong place to comprehensively document hooks; these types should be
at the point where they are defined instead.

I haven't added it there for now either though, because there were no
descriptions anyway so it wouldn't have been useful. Later we can
declare the functions properly where they are defined, and pass them
through in the place this map is assembled.

Gudahtt commented on app/scripts/lib/rpc-method-middleware/createMethodMiddleware.js:
```diff
@@ -1,5 +1,8 @@
+///: END:ONLY_INCLUDE_IN
 import { permissionRpcMethods } from '@metamask/snap-controllers';
-import { selectHooks } from '@metamask/rpc-methods';
+import { selectHooks } from '@metamask/rpc-methods/dist/utils';
```
This change is not needed - we don't need to include the path to the `utils` module

rekmarks replied:
Fixed in: 255d6033e

Gudahtt commented on lavamoat/build-system/policy-override.json:
```diff
@@ -30,8 +30,8 @@
       "packages": {
-        "eslint-import-resolver-node": true,
-        "@babel/eslint-parser": true
+        "@babel/eslint-parser": true,
```
This change seems unnecessary

Gudahtt commented on ui/components/app/connected-sites-list/connected-sites-list.component.js:
```diff
@@ -39,11 +39,12 @@ export default class ConnectedSitesList extends Component {
               onClick={() => onDisconnect(subject.origin)}
-            />
+            >
+              {t('disconnect')}
```
We could split out this change

Gudahtt replied:
Though also we shouldn't make clickable divs. This is better off as a button.

rekmarks replied:
Updated in #13381

We could perhaps still split it out @shanejonas

Gudahtt commented on ui/pages/confirmation/templates/index.js:
```diff
@@ -22,6 +28,7 @@ const ALLOWED_TEMPLATE_KEYS = [
   'cancelText',
   'onApprove',
   'onCancel',
+  'networkDisplay',
```
This is part of a change that makes the network display on these templates optional (because the snap confirmation didn't pertain to any network). We can split out this change.

rekmarks replied:
@hmalik88

Gudahtt commented on ui/pages/confirmation/confirmation.js:
```diff
@@ -185,23 +183,20 @@ export default function ConfirmationPage() {
+          </Box>
+        ) : null}
+        <Box justifyContent="center" padding={[4, 4, 4]}>
+          <SiteOrigin
```
We could split out this change to how the site origin is shown on templated pages

rekmarks replied:
@hmalik88

Nana12345678910 replied:
This is from my MetaMask account

Gudahtt commented on ui/hooks/useOriginMetadata.js:
```diff
@@ -19,24 +19,38 @@ import { SUBJECT_TYPES } from '../../shared/constants/app';
     };
   }
-  return minimumOriginMetadata;
+  return null;
```
Well. I don't feel great about this. I wonder if there's an alternative to using this hook for things that aren't actually origins :thinking:

rekmarks replied:
I think we need to take some time to consider how to deal with non-URL origins generally, maybe by moving to some generic notion of "subject identifiers" and let `origin` refer to URLs only and be optionally present.

Returning `null` here is more for the sake of completion however, because we should never end up in a situation where an "origin" doesn't have any metadata. Meanwhile, we could throw an error instead?

Gudahtt commented on app/_locales/en/messages.json:
```diff
@@ -2175,6 +2201,34 @@
+    "description": "The description for the `eth_accounts` permission"
+  },
+  "permission_manageBip44Keys": {
+    "message": "Manage your \"$1\" keys.",
```
Why was this text used rather than the one in the designs? That said "grant irrevocable control". This description does not convey the risks of granting this permission.

rekmarks replied:
This has been updated.

Gudahtt commented:
I have just rebased this on `develop` and squashed it down to a single commit again.

Gudahtt commented on app/scripts/background.js:
```diff
@@ -414,7 +414,10 @@ function setupController(initState, initLangCode) {
   function connectExternal(remotePort) {
     const portStream = new PortStream(remotePort);
-    controller.setupUntrustedCommunication(portStream, remotePort.sender);
+    controller.setupUntrustedCommunication({
```
This change could be split out

rekmarks replied:
@shanejonas

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -521,6 +547,72 @@ export default class MetamaskController extends EventEmitter {
+      getRpcMessageHandler: this.workerController.getRpcMessageHandler.bind(
+        this.workerController,
+      ),
+      // TODO:flask Fix in skunkworks (SnapController)
```
These two TODO comments, what needs to be fixed for them exactly?

rekmarks replied:
Using controller messenger actions instead of passing around functions. This has been done as of `skunkworks@0.8.1`.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -829,6 +927,49 @@ export default class MetamaskController extends EventEmitter {
+   * and Permission controllers during initialization.
+   */
+  getSnapPermissionSpecifications() {
+    // TODO:flask Probably find a better solution?
```
We have to do this in a few places already. I'm not sure there is a better solution.

Gudahtt replied:
Oh, right, these can be actions. Nevermind! We can update this comment to say that these should be actions.

rekmarks replied:
Updated to use actions as of `skunkworks@0.8.1`.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -2136,6 +2331,26 @@ export default class MetamaskController extends EventEmitter {
 
+  ///: BEGIN:ONLY_INCLUDE_IN(flask)
+  /**
+   * TODO:snaps verify safety of using keyringController.getAccounts
```
What was the concern here exactly?

rekmarks replied:
You know, I can't recall. However, I have concerns about deriving app keys from the user's Ethereum addresses (we should just use some arbitrary derivation path instead probably), so I updated the docstring to reflect that. Also renamed the method to `getAppKeyForSubject`.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -2882,6 +3165,35 @@ export default class MetamaskController extends EventEmitter {
+          this.snapController,
+          origin,
+        ),
+        // TODO:flask Fix in skunkworks (rpc-methods)
```
Same question here: fix what?

rekmarks replied:
I went through the commit history and can't find this comment anywhere. I recall fixing it, but not what the problem was. Whatever the case, it's not there anymore :man_shrugging:

Gudahtt commented on ui/components/app/flask/snap-settings-card/snap-settings-card.js:
```diff
@@ -55,36 +57,6 @@ const SnapSettingsCard = ({
   chipProps,
 }) => {
   const t = useI18nContext();
-  const [chipStatus, setChipStatus] = useState(STATUSES.INSTALLING);
```
We could split these changes out, since this component is on `develop` already

rekmarks replied:
@shanejonas

Gudahtt commented on ui/components/app/metamask-template-renderer/safe-component-list.js:
```diff
@@ -7,6 +7,7 @@ import Typography from '../../ui/typography';
 import Box from '../../ui/box';
 import MetaMaskTranslation from '../metamask-translation';
 import NetworkDisplay from '../network-display';
+import TextArea from '../../ui/textarea/textarea';
```
We could split this change out

rekmarks replied:
@hmalik88

rekmarks replied:
@shanejonas

Gudahtt commented on ui/components/app/multiple-notifications/index.scss:
```diff
@@ -56,8 +56,8 @@
     position: fixed;
-    bottom: 10px;
-    right: 10px;
+    bottom: 8px;
```
We can split this change out

rekmarks replied:
@shanejonas

Gudahtt commented on ui/components/app/permission-page-container/index.scss:
```diff
@@ -94,6 +94,7 @@
     display: flex;
     flex-direction: column;
     margin-top: 38px;
+    width: 100%;
```
Not entirely sure why this was needed :thinking: But if it is needed, we can split it out.

rekmarks replied:
@hmalik88

Gudahtt commented on ui/components/app/permissions-connect-footer/index.scss:
```diff
@@ -7,7 +7,7 @@
     @include H7;
 
-    color: #6a737d;
+    color: var(--ui-4);
```
This one too

rekmarks replied:
@hmalik88

Gudahtt commented on ui/components/app/permissions-connect-permission-list/index.scss:
```diff
@@ -1,8 +1,9 @@
 .permissions-connect-permission-list {
+  width: 100%;
```
Unsure why this was needed :thinking:

rekmarks replied:
@hmalik88

hmalik88 replied:
<img width="300" alt="Screen Shot 2022-01-26 at 2 23 20 PM" src="https://user-images.githubusercontent.com/41640681/151232145-79b928c4-4936-4cb7-b58d-cad93dae6ade.png">

hmalik88 replied:
Hmm reverted the change and everything was fine....unless I'm missing something @ritave?

Gudahtt commented on ui/components/app/permissions-connect-permission-list/permissions-connect-permission-list.js:
```diff
@@ -1,33 +1,129 @@
 
 PermissionsConnectPermissionList.propTypes = {
-  permissions: PropTypes.objectOf(PropTypes.bool).isRequired,
+  permissions: PropTypes.object.isRequired, // TODO(ritave): Currently some properties are boolean, some are {} empty objects, fix the issue at root cause.
```
I think they are supposed to be objects in all cases. At least according to the skunkworks types.

rekmarks replied:
Correct, but `getRequestedMethodState` in `permission-page-container.component` will return an object where all values are `true`. If we think that this is a problem we should fix the aforementioned method on `develop`. For now, I will remove this TODO.

Gudahtt commented on ui/components/ui/icon/verified-icon.js:
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+export default function IconVerified({
```
We can split this out as well

rekmarks replied:
Since we're not using it, I removed the file and all references to it from this branch.

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -2136,6 +2331,26 @@ export default class MetamaskController extends EventEmitter {
+   * @param domain
+   * @param requestedAccount
+   */
+  async getAppKeyForDomain(domain, requestedAccount) {
```
Consider removing this and adding it later with a proper permission

rekmarks replied:
During the initial discussions about app keys ~2 years ago, we never intended for them to require a permission. See also: https://github.com/MetaMask/metamask-extension/pull/11837#discussion_r790304000

rekmarks committed:
Update snap key access warning message

rekmarks committed:
Use latest skunkworks packages (#13299)

`skunkworks@0.7.0`. Snaps successfully tested locally.

rekmarks committed:
Update permission confirmation window title (#13342)

Adds a generic "Permission Request" title to the permission confirmation that's applied when `eth_accounts` is not requested. This does not address the case when both `eth_accounts` _and_ other permissions are requested, in which case we should arguably fall back to the generic title as well. Either way, this improves upon the status quo.

hmalik88 committed:
Snaps key access warning refactor (#13315)

* refactored snaps install warning

* removed unused import

* added fencing to flask components

* removed extraneous newlines

* removed popover css from SnapInstallWarning, added wrapper div to provide some padding

rekmarks committed:
Update key management permission description

rekmarks committed:
yarn-deduplicate

rekmarks committed:
Revert unnecessary import change

rekmarks committed:
Revert unnecessary change

rekmarks committed:
Clean up useOriginMetadata

rekmarks committed:
Remove useless comment

rekmarks committed:
yarn lavamoat:background:auto

rekmarks committed:
skunkworks@0.8.0

rekmarks committed:
skunkworks@0.8.1

rekmarks committed:
Improve app key getter docstring

rekmarks committed:
Remove todo comment

rekmarks committed:
Remove verified icon

rekmarks committed:
Use Button component for the disconnect button (#13381)

Use a `Button` component instead of a `div` in the connected sites list. This should look the same in the snap details page under settings and the connected sites popover from the three-dot menu in the popup home page.

hmalik88 committed:
Revert unnecessary CSS changes (#13429)

This PR reverts unnecessary CSS changes that were made in flask UI development.

rekmarks committed:
lint:fix

rekmarks commented:
Closing this PR in favor of a new one.
